llmops-article-classifier/steps/compare_models.py

```diff
@@ -0,0 +1,161 @@
+    claude_output_cost_per_1k = claude_haiku_token_costs["input_cost_per_1k"]
+
+    # Initialize models
+    tokenizer = AutoTokenizer.from_pretrained(tokenizer_path)
+    model = AutoModelForSequenceClassification.from_pretrained(modernbert_path)
+    model.eval()
+
+    # Get texts and labels
+    texts = [record["text"] for record in test_dataset]
+    labels = test_dataset["label"]
+    total_samples = len(texts)
+
+    # ModernBERT evaluation
+    with torch.no_grad():
+        all_logits = []
+        total_modernbert_latency = 0
+
+        for i in range(0, total_samples, modernbert_batch_size):
+            batch_texts = texts[i : i + modernbert_batch_size]
+            inputs = tokenizer(
+                batch_texts, padding=True, truncation=True, max_length=512, return_tensors="pt"
+            )
+
+            start_time = time.time()
+            outputs = model(**inputs)
+            batch_latency = time.time() - start_time
+
+            all_logits.extend(outputs.logits.numpy())
+            total_modernbert_latency += batch_latency
+
+    modernbert_latency = total_modernbert_latency / total_samples
+
+    # Claude evaluation
+    evaluator = ClaudeEvaluator(
+        batch_size=claude_batch_size,
+        api_key=anthropic_api_key,
+    )
+
+    # Create event loop for async operations
+    loop = asyncio.new_event_loop()
+    asyncio.set_event_loop(loop)
+    try:
+        claude_results = loop.run_until_complete(evaluator.evaluate_batch(texts))
+    finally:
+        loop.close()
+
+    # Filter valid results
+    valid_claude_results = [r for r in claude_results if r.error is None]
+    valid_indices = [i for i, r in enumerate(claude_results) if r.error is None]
+    valid_labels = np.array([labels[i] for i in valid_indices])
+
+    # Calculate Claude costs per prediction
+    claude_costs = []
+    for result in valid_claude_results:
+        input_cost = (result.input_tokens / 1000) * claude_input_cost_per_1k
+        output_cost = (result.output_tokens / 1000) * claude_output_cost_per_1k
+        total_cost = input_cost + output_cost
+        claude_costs.append(total_cost)
+
+    # Calculate average cost per prediction and scale to cost per 1000 predictions
+    avg_claude_cost = np.mean(claude_costs)
+    claude_cost_per_1000 = avg_claude_cost * 1000
+
+    # Prepare ModernBERT metrics
+    modernbert_logits = np.array([all_logits[i] for i in valid_indices])
+
+    # Prepare Claude metrics (with dummy logits for binary classification)
+    claude_predictions = np.array([r.prediction for r in valid_claude_results])
+    claude_logits = np.zeros((len(claude_predictions), 2))
+    claude_logits[range(len(claude_predictions)), claude_predictions] = 1
+
+    # Compute performance metrics
+    modernbert_performance = compute_classification_metrics((modernbert_logits, valid_labels))
+    claude_performance = compute_classification_metrics((claude_logits, valid_labels))
```

Comment:
In the future maybe I might extract some of this stuff out into helper functions. This starts to be quite a bit of code directly in the step. But no need to necessarily do that now.